Source/EmptyDefaultValueProvider.cs

```diff
@@ -73,73 +84,79 @@ public virtual object ProvideDefault(MethodInfo member)
+			else if (type.GetTypeInfo().IsGenericType)
 			{
-				// Task<T> inherits from Task, so just return Task<bool>
-				return GetCompletedTaskForType(typeof(bool));
+				factoryKey = type.GetGenericTypeDefinition();
 			}
-			else if (valueType.GetTypeInfo().IsGenericType && valueType.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+			else
 			{
-				var genericListType = typeof(List<>).MakeGenericType(valueType.GetGenericArguments()[0]);
-				return Activator.CreateInstance(genericListType);
+				factoryKey = type;
 			}
-			else if (valueType.GetTypeInfo().IsGenericType && valueType.GetGenericTypeDefinition() == typeof(IQueryable<>))
-			{
-				var genericType = valueType.GetGenericArguments()[0];
-				var genericListType = typeof(List<>).MakeGenericType(genericType);
 
-				return typeof(Queryable).GetMethods("AsQueryable")
-					.Single(x => x.IsGenericMethod)
-					.MakeGenericMethod(genericType)
-					.Invoke(null, new[] { Activator.CreateInstance(genericListType) });
-			}
-			else if (valueType.GetTypeInfo().IsGenericType && valueType.GetGenericTypeDefinition() == typeof(Task<>))
-			{
-				var genericType = valueType.GetGenericArguments()[0];
-				return GetCompletedTaskForType(genericType);
-			}
+			return factories.TryGetValue(factoryKey, out Func<Type, object> factory) ? factory.Invoke(type) : null;
+		}
 
-			return null;
+		private static object CreateArray(Type type)
+		{
+			var elementType = type.GetElementType();
+			var lengths = new int[type.GetArrayRank()];
+			return Array.CreateInstance(elementType, lengths);
 		}
 
-		private static object GetValueTypeDefault(Type valueType)
+		private static object CreateEnumerable(Type type)
 		{
-			// For nullable value types, return null.
-			if (valueType.GetTypeInfo().IsGenericType && valueType.GetGenericTypeDefinition() == typeof(Nullable<>))
-			{
-				return null;
-			}
+			return new object[0];
+		}
 
-			return Activator.CreateInstance(valueType);
+		private static object CreateEnumerableOf(Type type)
+		{
+			var elementType = type.GetGenericArguments()[0];
+			return Array.CreateInstance(elementType, 0);
+		}
+
+		private static object CreateQueryable(Type type)
+		{
+			return new object[0].AsQueryable();
+		}
+
+		private static object CreateQueryableOf(Type type)
+		{
+			var elementType = type.GetGenericArguments()[0];
+			var array = Array.CreateInstance(elementType, 0);
+
+			return typeof(Queryable).GetMethods("AsQueryable")
+				.Single(x => x.IsGenericMethod)
+				.MakeGenericMethod(elementType)
+				.Invoke(null, new[] { array });
 		}
 
-		private static Task GetCompletedTaskForType(Type type)
+		private static object CreateTask(Type type)
 		{
-			var tcs = Activator.CreateInstance(typeof (TaskCompletionSource<>).MakeGenericType(type));
+			return Task.FromResult(false);
+		}
 
-			var setResultMethod = tcs.GetType().GetMethod("SetResult");
-			var taskProperty = tcs.GetType().GetProperty("Task");
+		private static object CreateTaskOf(Type type)
```

Comment:
Interestingly enough, while creating a completed task via `Task.FromResult<T>` is about 16x as fast as via `TaskCompletionSource<T>`, the same is not true if done via reflection: `Task.FromResult<T>` then becomes a little slower even though less reflection calls are involved. So we don't change over to a `Task.FromResult`-based solution.